Feature - adding better auth for handling authentication and session management (server (new PR))

PurnenduMIshra129th opened the pull request:
What kind of change does this PR introduce?

feature

Issue Number: #3305 

Fixes

Snapshots/Videos:
![image](https://github.com/user-attachments/assets/d4dd8077-4e49-4678-bfc0-ce74656a17b6)

If relevant, did you update the documentation?

Yes i have updated the documentation in docs/docs/docs/developer-resources

Summary
-Now Login , sign up, sign out, session management , and enable cookie (similar to refresh token) is implemented with better auth.

Does this PR introduce a breaking change?

-Yes in this pr we have to migrate the database schema.due to addition of session , account, verification table in database.

## Checklist

### CodeRabbit AI Review
- [x] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [x] I have implemented or provided justification for each non-critical suggestion
- [x] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass


**Other information**
It is related to our new authentication system.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**
	- Enhanced the API service configuration by introducing two required environment variables: `API_CORS_ORIGIN` and `BETTER_AUTH_SECRET`, ensuring necessary settings are provided during deployment.
	- Updated the `.gitignore` to modify the ignore rule for `gql.tada.d.ts`, allowing it to apply to any subdirectory within the repository.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

PurnenduMIshra129th commented on src/drizzle/tables/verification.ts:
```diff
@@ -0,0 +1,9 @@
+	expiresAt: timestamp("expires_at").notNull(),
+	createdAt: timestamp("created_at"),
+	updatedAt: timestamp("updated_at"),
+});
```
documentation is done in docs/docs/docs/developer-resources/better-auth.md

PurnenduMIshra129th commented on src/drizzle/tables/session.ts:
```diff
@@ -0,0 +1,14 @@
+	id: uuid("id").primaryKey().defaultRandom(),
+	expiresAt: timestamp("expires_at").notNull(),
+	token: text("token").notNull().unique(),
+	createdAt: timestamp("created_at").notNull(),
+	updatedAt: timestamp("updated_at").notNull(),
```
not required here it is auto generating in inserting the record so it is fine.

PurnenduMIshra129th commented on src/drizzle/tables/session.ts:
```diff
@@ -0,0 +1,14 @@
+	userId: uuid("user_id")
+		.notNull()
+		.references(() => usersTable.id, { onDelete: "cascade" }),
+});
```
not required to do any modification in default table of better auth as it is auto generated during schema generation.

PurnenduMIshra129th commented on src/drizzle/tables/account.ts:
```diff
@@ -0,0 +1,20 @@
+	password: text("password"),
+	createdAt: timestamp("created_at").notNull(),
+	updatedAt: timestamp("updated_at").notNull(),
+});
```
not required here also it is default table we don't need to change anything here

PurnenduMIshra129th commented on drizzle_migrations/20250331213202_pale_shiver_man.sql:
```diff
@@ -0,0 +1,44 @@
+CREATE TABLE "account" (
+	"id" uuid PRIMARY KEY DEFAULT gen_random_uuid() NOT NULL,
+	"account_id" text NOT NULL,
+	"provider_id" text NOT NULL,
+	"user_id" uuid NOT NULL,
+	"access_token" text,
+	"refresh_token" text,
+	"id_token" text,
+	"access_token_expires_at" timestamp,
+	"refresh_token_expires_at" timestamp,
+	"scope" text,
+	"password" text,
+	"created_at" timestamp NOT NULL,
+	"updated_at" timestamp NOT NULL
+);
```
auto generated schema by orm  no need to change here.

PurnenduMIshra129th commented on drizzle_migrations/20250331213202_pale_shiver_man.sql:
```diff
@@ -0,0 +1,44 @@
+	"updated_at" timestamp NOT NULL
+);
+--> statement-breakpoint
+CREATE TABLE "session" (
+	"id" uuid PRIMARY KEY DEFAULT gen_random_uuid() NOT NULL,
+	"expires_at" timestamp NOT NULL,
+	"token" text NOT NULL,
+	"created_at" timestamp NOT NULL,
+	"updated_at" timestamp NOT NULL,
+	"ip_address" text,
+	"user_agent" text,
+	"user_id" uuid NOT NULL,
+	CONSTRAINT "session_token_unique" UNIQUE("token")
+);
```
auto generated schema by adapter no need to change any thing.

PurnenduMIshra129th commented on src/graphql/types/Query/signIn.ts:
```diff
@@ -72,7 +72,8 @@ builder.queryField("signIn", (t) =>
-				!(await verify(existingUser.passwordHash, parsedArgs.input.password))
+				!(await verify(
+					existingUser.passwordHash ? existingUser.passwordHash : '', parsedArgs.input.password))
 			) {
```
this component is not required because we are using better auth for authentication rather than custom logic.

PurnenduMIshra129th commented on src/drizzle/tables/account.ts:
```diff
@@ -0,0 +1,20 @@
+	accessTokenExpiresAt: timestamp("access_token_expires_at"),
+	refreshTokenExpiresAt: timestamp("refresh_token_expires_at"),
+	scope: text("scope"),
+	password: text("password"),
```
not required for now

PurnenduMIshra129th commented on src/lib/auth.ts:
```diff
@@ -0,0 +1,118 @@
+			maxAge: 5 * 60, // Cache duration in seconds
+		},
+	},
+	emailAndPassword: {
+		enabled: true,
+	},
+	hooks: {
+		after: createAuthMiddleware(async (ctx) => {
+			if (ctx.path.startsWith("/sign-up")) {
+				try {
+					const newSession = ctx.context.newSession;
+					if (newSession) {
+						const user = newSession.user;
+						const session = newSession.session;
+
+						return ctx.json({
+							statusCode: "10000",
+							message: "Success",
+							data: {
+								token: session.token,
+								id: user.id,
+							},
+						});
+					}
+				} catch (error: unknown) {
+					console.log(error);
+					return {
+						statusCode: "10001",
+						message: "Failure",
+						error: error instanceof Error ? error.message : "Unknown error",
+					};
+				}
+			}
```
this can be adjusted in future when required

PurnenduMIshra129th commented on src/lib/auth.ts:
```diff
@@ -0,0 +1,118 @@
+import { betterAuth } from "better-auth";
+import { drizzleAdapter } from "better-auth/adapters/drizzle";
+import { createAuthMiddleware } from "better-auth/api";
+import dotenv from "dotenv";
+import { eq } from "drizzle-orm";
+import {
+	accountTable,
+	sessionTable,
+	usersTable,
+	verificationTable,
+} from "../drizzle/schema";
+import { db } from "./db";
+dotenv.config();
+
```
now i have added trusted origin code so dotenv is requried

PurnenduMIshra129th commented on src/lib/db.ts:
```diff
@@ -0,0 +1,15 @@
+	// 	console.log("Running SQL Query:", query);
+	// 	console.log("📌 Query Parameters:", params);
+	// },
+});
```
not required for now . It is working fine

PurnenduMIshra129th commented:
now failed test case is solved  in server only few   suggestion i have to take that's it .

PurnenduMIshra129th commented:
now server is completed just have to add the 
API_CORS_ORIGIN = http://localhost:4321
NODE_ENV=development
in the env and test will not fall is it my part ? becasue my env is in my local and added validation for other user in our existing validation schema ? do i have to change any thing?

palisadoes commented:
1. Make the required modifications to all relevant files
2. You will also need to update the setup script

PurnenduMIshra129th commented:
@palisadoes can you tell which changes you need to be present so that it can be merged

palisadoes commented:
You will also need to update this file:
```
docs/docs/docs/getting-started/environment-variables.md
```

PurnenduMIshra129th commented:
now i have added the documentation of env in docs update alll the instance of env and added in setup script so test cases is passed but i don't know which files i do not changed also automated documentation is generating docs what is the reason ? that's why one test case is falling . and the test cases i will write for create server.ts so it will match the expected value . i am working on this test cases percentage.

PurnenduMIshra129th commented:
@palisadoes  in codecov report it is showing some lines are missing in test cases but for createServer i did not find any test file like createServer.spec.ts or createServer.test.ts so where should i write the test cases ? all changes are done except test cases in this file and in signup also it is showing some missing line but is it required cover that in our test cases as it is not used anymore. currently i have added setup script for better auth.

PurnenduMIshra129th commented:
it becomes complex to handle conlicts in auto generated files editor also not supporting  sometimes it closes automatically . as the day passes the auto generated files are coming in > 400 or >500 so it becomes difficult to manage so i am forcefully closes this ? i waited for 20 min but in merge editor the conflicts file did not open . we have to close this soon reviewers .
